src/core/tiledmesh/qgscesiumtilesdataprovider.cpp

```diff
@@ -28,26 +28,430 @@
+    {
+      tile->setResources( {{ QStringLiteral( "content" ), contentUri } } );
+    }
+  }
+
+  return tile;
+}
+
+QgsTiledMeshNode *QgsCesiumTiledMeshIndex::nodeFromJson( const json &json, QgsTiledMeshNode *parent )
+{
+  std::unique_ptr< QgsTiledMeshTile > tile = tileFromJson( json, parent ? parent->tile() : nullptr );
+  std::unique_ptr< QgsTiledMeshNode > newNode = std::make_unique< QgsTiledMeshNode >( tile.release() );
+  mNodeMap.insert( newNode->tile()->id(), newNode.get() );
+
+  if ( parent )
+    parent->addChild( newNode.get() );
+
+  if ( json.contains( "children" ) )
+  {
+    for ( const auto &childJson : json["children"] )
+    {
+      nodeFromJson( childJson, newNode.get() );
+    }
+  }
+
+  return newNode.release();
+}
+
+void QgsCesiumTiledMeshIndex::refineNodeFromJson( QgsTiledMeshNode *node, const json &json )
+{
+  std::unique_ptr< QgsTiledMeshTile > newTile = tileFromJson( json, node->parentNode() ? node->parentNode()->tile() : nullptr );
+  // copy just the resources from the retrieved tileset to the refined node. We assume all the rest of the tile content
+  // should be the same between the node being refined and the root node of the fetched sub dataset!
+  // (Ie the bounding volume, geometric error, etc).
+  node->tile()->setResources( newTile->resources() );
+
+  if ( json.contains( "children" ) )
+  {
+    for ( const auto &childJson : json["children"] )
+    {
+      nodeFromJson( childJson, node );
+    }
+  }
+}
+
+QgsTiledMeshTile QgsCesiumTiledMeshIndex::rootTile() const
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  return mRootNode ? *mRootNode->tile() : QgsTiledMeshTile();
+}
+
+QgsTiledMeshTile QgsCesiumTiledMeshIndex::getTile( const QString &id )
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  auto it = mNodeMap.constFind( id );
+  if ( it != mNodeMap.constEnd() )
+  {
+    return *( it.value()->tile() );
+  }
+
+  return QgsTiledMeshTile();
+}
+
+QString QgsCesiumTiledMeshIndex::parentTileId( const QString &id ) const
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  auto it = mNodeMap.constFind( id );
+  if ( it != mNodeMap.constEnd() )
+  {
+    if ( QgsTiledMeshNode *parent = it.value()->parentNode() )
+    {
+      return parent->tile()->id();
+    }
+  }
+
+  return QString();
+}
+
+QStringList QgsCesiumTiledMeshIndex::childTileIds( const QString &id ) const
+{
+  QgsReadWriteLocker locker( mLock, QgsReadWriteLocker::Read );
+  auto it = mNodeMap.constFind( id );
+  if ( it != mNodeMap.constEnd() )
+  {
+    QStringList childIds;
+    const QList< QgsTiledMeshNode * > children = it.value()->children();
+    childIds.reserve( children.size() );
+    for ( QgsTiledMeshNode *child : children )
+    {
+      childIds << child->tile()->id();
+    }
+    return childIds;
+  }
+
+  return QStringList();
+}
+
+QStringList QgsCesiumTiledMeshIndex::getTiles( const QgsTiledMeshRequest &request, const QString &parentId )
```

Comment:
getTiles() implementation will get the tileCanRefine() + refineAsync() support later I guess?

Reply:
@wonder-sk I am thinking it will be used like this: https://github.com/nyalldawson/QGIS/blob/tiled_mesh_renderer2/src/core/tiledmesh/qgstiledmeshlayerrenderer.cpp#L644-L680

Reply:
I see - but wouldn't it be better to integrate this into getTiles()? Otherwise we would need to implement the same logic in various places where getTiles() may be needed - identify, profile tool, ...

Reply:
The issue is that we'd then need two versions of getTiles -- one which is allowed to block during network requests, and the other which somehow allows the event loop to keep processing while requests are handled..

Reply:
Ideally, yes. But for the time being, we could live with just one version - the one that blocks - and possibly implement non-blocking variant later. I think users are quite used to the fact that some operations (like identify) can block the GUI :wink: 

Some more arguments for having the logic in getTiles() :slightly_smiling_face: 
- plugin developers using the getTiles() API it may be also difficult to wrap their heads around how to implement the logic... It would be very convenient if they could just use a high-level API to get everything needed and not worry about the details of populating the hierarchy.
- it would be nice if getTiles() would always return the same result irrespective of what bits of hierarchy were already seen in the given index
- getTiles() could optimize network requests - e.g. spawn multiple requests at once and wait for all of them - I assume the logic on the client side would most likely serialize the requests for simplicity

Reply:
Ok -- I'm gonna go for the most flexible option and add flags to QgsTiledMeshRequest. By default we'll fetch the hierarchy in a blocking call, but the flags allow for a caller to opt out of this if they are going to handling the hierarchy fetch/refinement themselves.